src/actions/assetsActions.js

```diff
@@ -502,18 +511,48 @@ export const fetchAllAccountsTotalBalancesAction = () => {
             // no need to add anything no matching account found or empty balances
             if (!account || isEmpty(accountProtocolBalances)) return;
 
-            const accountCategoryBalance = accountProtocolBalances.reduce((categoryBalance, balances) => {
-              const protocolBalances = balances?.assets;
-              if (isEmpty(protocolBalances)) return categoryBalance;
-
-              let protocolBalanceValues = protocolBalances.map(({ balanceUSD }) => BigNumber(balanceUSD ?? 0));
-
-              if (currency !== USD) {
-                const rate = zapperUSDBasedRates?.[currency.toUpperCase()];
-                protocolBalanceValues = protocolBalanceValues.map((value) => value.times(rate ?? 0));
-              }
-
-              return sum([categoryBalance, ...protocolBalanceValues]);
+            let categoryAssetsBalances = [];
+
+            const categoryTotalBalance = accountProtocolBalances.reduce((categoryBalance, balances) => {
+              const protocolAssetsBalances = balances?.assets;
+              if (isEmpty(protocolAssetsBalances)) return categoryBalance;
+
+              const assetsBalances = protocolAssetsBalances.map((asset) => {
+                const {
+                  protocol,
+                  symbol,
+                  label,
+                  balanceUSD,
+                  img,
+                  share: tokensShare,
+                  supply: tokensSupply,
+                } = asset;
+
+                const balanceUsdBN = BigNumber(balanceUSD);
+                const value = currency === USD
+                  ? balanceUsdBN
+                  : balanceUsdBN.times(usdRate ?? 0);
+
+                // TODO: do we need to fix exponential for very small values or hide small share at all?
+                const share = tokensShare && tokensSupply
+                  ? BigNumber(tokensShare).dividedBy(tokensSupply).times(100)
+                  : null;
+
+                return {
+                  key: `${protocol}-${symbol}`,
+                  service: balances.label,
+                  title: label,
+                  iconUrl: img ? `https://zapper.fi/images/${img}` : null,
+                  share,
+                  value,
+                };
+              });
+
+              categoryAssetsBalances = [...categoryAssetsBalances, ...assetsBalances];
```

Comment:
Seems like this reduce is not pure function because it calculating both total & asset array, this can be confusing to discover. Consider having adding 
```
let categoryTotalBalances = BigNumber(0) 
```
next to `let categoryAssetsBalances = [];`

and replacing `reduce` with `forEach`.

Reply:
Yeah I guess either way is fine, but putting totals outside the loop makes more sense 👍